src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs

```diff
@@ -0,0 +1,247 @@
+using System;
+using System.Linq;
+using Draco.Compiler.Internal.Query;
+using Draco.Compiler.Internal.Semantics.Symbols;
+using Draco.Compiler.Api.Syntax;
+using System.Collections.Immutable;
+using Type = Draco.Compiler.Internal.Semantics.Types.Type;
+
+namespace Draco.Compiler.Internal.Semantics.AbstractSyntax;
+
+/// <summary>
+/// Computations for building the AST.
+/// </summary>
+internal static class AstBuilder
+{
+    /// <summary>
+    /// Builds an <see cref="Ast"/> from the given <see cref="ParseTree"/>.
+    /// </summary>
+    /// <param name="db">The <see cref="QueryDatabase"/> for the computation.</param>
+    /// <param name="ast">The <see cref="ParseTree"/> to construct the <see cref="Ast"/> from.</param>
+    /// <returns>The <see cref="Ast"/> form of <paramref name="ast"/>.</returns>
+    public static Ast ToAst(QueryDatabase db, ParseTree ast) => ast switch
+    {
+        ParseTree.CompilationUnit cu => ToAst(db, cu),
+        ParseTree.Decl decl => ToAst(db, decl),
+        ParseTree.Stmt stmt => ToAst(db, stmt),
+        ParseTree.Expr expr => ToAst(db, expr),
+        _ => throw new ArgumentOutOfRangeException(nameof(ast)),
+    };
+
+    /// <summary>
+    /// Builds an <see cref="Ast"/> from the given <see cref="ParseTree"/>.
+    /// </summary>
+    /// <param name="db">The <see cref="QueryDatabase"/> for the computation.</param>
+    /// <param name="decl">The <see cref="ParseTree"/> to construct the <see cref="Ast"/> from.</param>
+    /// <returns>The <see cref="Ast"/> form of <paramref name="decl"/>.</returns>
+    public static Ast.Decl ToAst(QueryDatabase db, ParseTree.Decl decl) => db.GetOrUpdate(
+        decl,
+        Ast.Decl (decl) => decl switch
+        {
+            // TODO: Eliminate the ?? pattern everywhere by making the API use optional
+            // TODO: Eliminate the null ? null : ... pattern everywhere by making the API use optional
+
+            ParseTree.Decl.Func func => new Ast.Decl.Func(
+                ParseTree: func,
+                DeclarationSymbol: (ISymbol.IFunction?)SymbolResolution.GetDefinedSymbolOrNull(db, func) ?? throw new InvalidOperationException(),
+                Body: ToAst(db, func.Body)),
+            ParseTree.Decl.Label label => new Ast.Decl.Label(
+                ParseTree: label,
+                LabelSymbol: (ISymbol.ILabel?)SymbolResolution.GetDefinedSymbolOrNull(db, label) ?? throw new InvalidOperationException()),
+            ParseTree.Decl.Variable var => new Ast.Decl.Variable(
+                ParseTree: var,
+                DeclarationSymbol: (ISymbol.IVariable?)SymbolResolution.GetDefinedSymbolOrNull(db, var) ?? throw new InvalidOperationException(),
+                Value: var.Initializer is null ? null : ToAst(db, var.Initializer.Value)),
```

Comment:
We could really just have a `SymbolResolution.GetDefinedSymbolExpected<TSymbol>` in the API